src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_ObjectOrCollectionInitializerExpression.cs

```diff
@@ -156,6 +156,14 @@ private BoundExpression MakeCollectionInitializer(BoundExpression rewrittenRecei
             // such as generating a params array, re-ordering arguments based on argsToParamsOpt map, inserting arguments for optional parameters, etc.
             ImmutableArray<LocalSymbol> temps;
             var argumentRefKindsOpt = default(ImmutableArray<RefKind>);
+            if (addMethod.Parameters[0].RefKind == RefKind.Ref)
+            {
+                // If the Add method is an extension which takes a `ref this` as the first parameter, implicitly add a `ref` to the argument
+                var builder = ArrayBuilder<RefKind>.GetInstance(addMethod.Parameters.Length, RefKind.None);
```

Comment:
What happens when one of the other parameters is being passed by `ref`, `in`, etc ... Will this array eventually get rewritten to have the proper values? Or will the use of `RefKind.None` override the user code?

Reply:
`ref`, `in`, and `out` keywords are not allowed in collection initializer element syntax. Therefore the only RefKinds we'll find for the subsequent arguments are `RefKind.None`.

If the Add method's subsequent parameters use `in`, the generated Add call's argument's RefKind will be updated to `RefKind.In` in `LocalRewriter.BuildStoresToTemps`.

Reply:
Consider adding an assert that the remaining parameters are either `None` or `In`, and a comment that `In` will be updated as you mentioned.

---
In reply to: [224926176](https://github.com/dotnet/roslyn/pull/30481#discussion_r224926176) [](ancestors = 224926176)

Reply:
Added in 23085c7